Reduce memory usage when returning large result sets

Azquelt opened the pull request:
When returning large resultsets, stream assets one by one from the database to the response without holding the entire response in memory.

Azquelt committed:
Reduce memory usage when returning large result sets

This change allows results to be streamed from the database into a
response to the client when retrieving multiple assets. It avoids
needing to load the whole result set into memory before returning it.

To do this we introduce the AssetCursor, which wraps a database result
cursor and allows the Rest layer to iterate through the results without
needing to know the details of the persistence backend.

To allow changes to be made to the information from the database before
it is consumed by the rest layer we introduce the AssetOperation, which
encapsulates some processing which should be done to the asset before it
is returned (e.g. transforming between data types or removing
implementation-specific fields)

Azquelt committed:
Fix ordering of expected and actual arguments

Also includes a tiny bit of reformatting

Azquelt committed:
Move MongoDB support classes to their own package

idlewis commented on server/src/main/java/com/ibm/ws/lars/rest/PersistenceBean.java:
```diff
@@ -262,27 +269,24 @@ private BasicDBObject createFilterObject(String field, Condition condition) {
-        return results;
+
+        AssetCursor result = new MongoAssetCursor(cursor);
+        result.addOperation(CONVERT_OID_TO_HEX);
```
Both places we currently the new MongoAssetCursor, we add the convert operation. would it make sense to have a factory for this? MongoAssetCursor.newCursorWithIdConversion(cursor) or similar?

Azquelt replied:
Hmm, possibly? Or I could put this functionality into `MongoAssetCursor` itself on the basis that we will always need to convert the id when turning a MongoDB record into an asset.

On the other hand, it fits quite nicely into an operation.

Azquelt committed:
Update copyright year for new files

idlewis commented on server/src/test/java/com/ibm/ws/lars/rest/BasicAssetCursor.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Override
+    public void remove() {
```
Don't think there is any use-case for this, but isn't 'remove' supported by the map iterator?

Azquelt replied:
It is, but it's not supported in the MongoAssetCursor, so there doesn't seem much point implementing it in the test implementation

idlewis commented on server/src/main/java/com/ibm/ws/lars/rest/model/AssetCursor.java:
```diff
@@ -0,0 +1,52 @@
+     *
+     * @return the number of assets
+     */
+    public int getCount();
```
any reason why you called this getCount? It seems to replace a method call to 'size' in lots of places, and wraps a call to 'size' in the case of the mongo cursor.

Azquelt replied:
Not really, `getCount()` made sense to me when writing it, but you're correct that other objects use `size()`

Azquelt replied:
For added confusion, `DBCursor` has `length()`, `count()` and `size()` which all do different things

Azquelt committed:
Move AssetList into test src folder

With the introduction of AssetCursor, AssetList is now only used for
tests.

idlewis commented on server/src/main/java/com/ibm/ws/lars/rest/AssetCursorWriter.java:
```diff
@@ -0,0 +1,58 @@
+
+    /** {@inheritDoc} */
+    @Override
+    public void writeTo(AssetCursor cursor, Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType, MultivaluedMap<String, Object> httpHeaders,
```
You're not checking the mediatype here. At this point, do we just have to assume that jax-rs has called the above method and is giving us something sensible?

Azquelt replied:
Yeah, that was what I was assuming

Azquelt committed:
Ensure AssetCursors are closed when written

Make AssetCursor implement Closable, and then read the AssetCursor
within a try-with-resources so that it gets closed.

Azquelt committed:
Reformat RepositoryRESTResource

Newer version of Eclipse wants to reformat this class when I save, so
let's separate the reformatting changes into another commit.

Azquelt commented:
I checked and we do have a check to ensure the search match score is not returned in the response. It's at the end of `PersistenceBeanTest.testRetrieveAllFiltered()`

idlewis reviewed (approved):
Ok
